library/gui/screen/src/main/java/org/quiltmc/qsl/screen/api/client/ScreenEvents.java

```diff
@@ -0,0 +1,210 @@
+import org.quiltmc.qsl.base.api.event.client.ClientEventAwareListener;
+import org.quiltmc.qsl.screen.impl.client.ScreenExtensions;
+
+import net.minecraft.client.MinecraftClient;
+import net.minecraft.client.gui.screen.Screen;
+import net.minecraft.client.util.math.MatrixStack;
+
+import net.fabricmc.api.EnvType;
+import net.fabricmc.api.Environment;
+
+/**
+ * Holds events related to {@link Screen}s.
+ *
+ * <p>Some events require a screen instance in order to obtain an event instance.
+ * The events that require a screen instance can be identified by the use of a method passing a screen instance.
+ * All events in {@link ScreenKeyboardEvents} and {@link ScreenMouseEvents} require a screen instance.
+ * This registration model is used since a screen being (re)initialized will reset the screen to it's default state, therefore reverting all changes a mod developer may have applied to a screen.
+ * Furthermore this design was chosen to reduce the amount of wasted iterations of events as a mod developer would only need to register screen events for rendering, ticking, keyboards and mice if needed on a per instance basis.
+ *
+ * <p>The primary entrypoint into a screen is when it is being opened, this is signified by an event {@link ScreenEvents#BEFORE_INIT before} and {@link ScreenEvents#AFTER_INIT after} initialization of the screen.
+ *
+ * @see QuiltScreenHooks
+ * @see ScreenKeyboardEvents
+ * @see ScreenMouseEvents
+ */
+@Environment(EnvType.CLIENT)
+public final class ScreenEvents {
+	/**
+	 * An event that is called before {@link Screen#init(MinecraftClient, int, int) a screen is initialized} to it's default state.
+	 * It should be noted some of the methods in {@link QuiltScreenHooks} such as a screen's {@link QuiltScreenHooks#getTextRenderer(Screen) text renderer} may not be initialized yet, and as such their use is discouraged.
+	 *
+	 * <!--<p>Typically this event is used to register screen events such as listening to when child elements are added to the screen. ------ Uncomment when child add/remove event is added for elements-->
+	 * You can still use {@link ScreenEvents#AFTER_INIT} to register events such as keyboard and mouse events.
+	 *
+	 * <p>The {@link ScreenExtensions} provided by the {@code info} parameter may be used to register tick, render events, keyboard, mouse, additional and removal of child elements (including buttons).
+	 * For example, to register an event on inventory like screens after render, the following code could be used:
+	 * <pre>{@code
+	 * &#64;Override
+	 * public void onInitializeClient() {
+	 * 	ScreenEvents.BEFORE_INIT.register((client, screen, scaledWidth, scaledHeight) -> {
+	 * 		if (screen instanceof AbstractInventoryScreen) {
+	 * 			ScreenEvents.getAfterRenderEvent(screen).register((matrices, mouseX, mouseY, tickDelta) -> {
+	 * 				...
+	 * 			});
+	 * 		}
+	 * 	});
+	 * }
+	 * }</pre>
+	 *
+	 * <p>This event indicates a screen has been resized, and therefore is being re-initialized.
```

Comment:
Information explaining what the event does should be above the example

Reply:
Fixed.